plutus-use-cases/src/Plutus/Contracts/Uniswap.hs

```diff
@@ -0,0 +1,988 @@
+-- Coins are the entities that can be swapped in the exchange.
+data Coin = Coin
+    { cCurrency :: CurrencySymbol
+    , cToken    :: TokenName
+    } deriving (Show, Generic, ToJSON, FromJSON, ToSchema)
+
+PlutusTx.unstableMakeIsData ''Coin
+PlutusTx.makeLift ''Coin
+
+instance Eq Coin where
+    {-# INLINABLE (==) #-}
+    c == d = cCurrency c == cCurrency d && cToken c == cToken d
+
+instance Prelude.Eq Coin where
+    (==) = (==)
+
+{-# INLINABLE compareCoins #-}
+compareCoins :: Coin -> Coin -> Ordering
+compareCoins c d = case compare (cCurrency c) (cCurrency d) of
+    LT -> LT
+    GT -> GT
+    EQ -> compare (cToken c) (cToken d)
+
+instance Prelude.Ord Coin where
+    compare = compareCoins
+
+{-# INLINABLE coinLT #-}
+coinLT :: Coin -> Coin -> Bool
+coinLT c d = case compareCoins c d of
+    LT -> True
+    _  -> False
+
+{-# INLINABLE coin #-}
+-- | @'coin' c n@ denotes the value given by @n@ units of @'Coin'@ @c@.
+coin :: Coin    -- ^ The 'Coin'.
+     -> Integer -- ^ The desired number coins.
+     -> Value   -- ^ The 'Value' consisting of the given number of units of the given 'Coin'.
+coin Coin{..} = Value.singleton cCurrency cToken
+
+{-# INLINABLE coinValueOf #-}
+-- | Calculates how many units of the specified 'Coin' are contained in the
+-- given 'Value'.
+coinValueOf :: Value   -- ^ The 'Value' to inspect.
+            -> Coin    -- ^ The 'Coin' to look for.
+            -> Integer -- ^ The number of units of the given 'Coin' contained in the given 'Value'.
+coinValueOf v Coin{..} = valueOf v cCurrency cToken
+
+{-# INLINABLE hashCoin #-}
+hashCoin :: Coin -> ByteString
+hashCoin Coin{..} = sha2_256 $ concatenate (unCurrencySymbol cCurrency) (unTokenName cToken)
+
+data Sqrt =
+      Imaginary
+    | Exact Integer
+    | Irrational Integer
+    deriving stock Show
+
+PlutusTx.unstableMakeIsData ''Sqrt
+PlutusTx.makeLift ''Sqrt
+
+{-# INLINABLE rsqrt #-}
+rsqrt :: Integer -> Integer -> Sqrt
+rsqrt n d
+    | n * d < 0 = Imaginary
+    | n == 0    = Exact 0
+    | n == d    = Exact 1
+    | n < 0     = rsqrt (negate n) (negate d)
+    | otherwise = go 1 $ 1 + divide n d
+  where
+    go :: Integer -> Integer -> Sqrt
+    go l u
+        | l * l * d == n = Exact l
+        | u == (l + 1)   = Irrational l
+        | otherwise      =
+              let
+                m = divide (l + u) 2
+              in
+                if m * m * d <= n then go m u
+                                  else go l m
+
+{-# INLINABLE isqrt #-}
+isqrt :: Integer -> Sqrt
+isqrt n = rsqrt n 1
+
+{-# INLINABLE calculateInitialLiquidity #-}
+calculateInitialLiquidity :: Integer -> Integer -> Integer
+calculateInitialLiquidity outA outB = case isqrt (outA * outB) of
+    Exact l
+        | l > 0 -> l
+    Irrational l
+        | l > 0 -> l + 1
+    _           -> traceError "insufficient liquidity"
+
+{-# INLINABLE calculateAdditionalLiquidity #-}
+calculateAdditionalLiquidity :: Integer -> Integer -> Integer -> Integer -> Integer -> Integer
```

Comment:
IMO this module could do with some newtypes over the various `Integer` quantities, or at least type synonyms.